jax/_src/cudnn/fused_attention_stablehlo.py

```diff
@@ -0,0 +1,728 @@
+    is_causal_mask=is_causal_mask)
+  
+  # reshape to original shape
+  grad_query = jnp.reshape(grad_query, (*batch_tuple, q_seq_len, num_heads, head_dim))
+  grad_key = jnp.reshape(grad_key, (*batch_tuple, kv_seq_len, num_heads, head_dim))
+  grad_value = jnp.reshape(grad_value, (*batch_tuple, kv_seq_len, num_heads, head_dim))
+  grads = (grad_query, grad_key, grad_value)
+  return grads, out_bdims
+
+# custom partitioning
+def _get_padded_spec(arg_info):
+  spec = None if arg_info.sharding is None else arg_info.sharding.spec
+  ndim = arg_info.ndim
+  if spec is None:
+    return (None,) * ndim
+  assert len(spec) <= ndim
+  return spec + (None,) * (ndim - len(spec))
+
+# fwd custom partition
+def _infer_fwd_output_sharding(mesh, arg_shapes):
+  # only sharding on batch and num_head dim is allowed
+  # (*batch, q_seq, num_head, head)
+  query_spec = _get_padded_spec(arg_shapes[0])
+  # (*batch, kv_seq, num_head, head)
+  key_spec = _get_padded_spec(arg_shapes[1])
+  value_spec = _get_padded_spec(arg_shapes[2])
+  if not query_spec == key_spec == value_spec:
+    raise ValueError("Query, key and value should have same sharding.")
+  seq_spec = query_spec[-3]
+  head_spec = query_spec[-1]
+  if seq_spec != None:
+    raise ValueError("Sharding on sequence dim is not allowed.")
+  if head_spec != None:
+    raise ValueError("Sharding on head dim is not allowed.")
+  # keep out sharding same as query sharding since they have same shape
+  out_sharding = NamedSharding(mesh, PartitionSpec(*query_spec))
+  # activation sharding
+  activation_sharding = NamedSharding(mesh, PartitionSpec(*query_spec[:-3], query_spec[-2], query_spec[-3], None))
+  return (out_sharding, activation_sharding)
+
+_dot_product_attention_fwd_lower = custom_partitioning(_dot_product_attention_fwd_impl, static_argnums=(5,6,7,8,9,10))
+def _dot_product_attention_fwd_infer_sharding_from_operands(scale, seed, dropout_rate, variadic_args, is_flash_attention, is_causal_mask, mesh, arg_shapes, result_shape):
+  return _infer_fwd_output_sharding(mesh, arg_shapes)
+
+def _dot_product_attention_fwd_partition(scale, seed, dropout_rate, variadic_args, is_flash_attention, is_causal_mask, mesh, arg_shapes, result_shape):
+  # args sharding
+  arg_shardings = tuple([arg_i.sharding for arg_i in arg_shapes])
+  out_shardings = _infer_fwd_output_sharding(mesh, arg_shapes)
+  impl = partial(_dot_product_attention_fwd_impl, scale=scale, seed=seed, dropout_rate=dropout_rate,
+                variadic_args=variadic_args, is_flash_attention=is_flash_attention, is_causal_mask=is_causal_mask)
+  return mesh, impl, out_shardings, arg_shardings
+
+# bwd custom partition
+def _infer_bwd_output_sharding(mesh, arg_shapes):
+  # (*batch, q_seq, num_head, head)
+  query_spec = _get_padded_spec(arg_shapes[0])
+  # (*batch, kv_seq, num_head, head)
+  key_spec = _get_padded_spec(arg_shapes[1])
+  # keep grad query sharding same as query sharding
```

Comment:
Here you need to raise an error if it tries to shard on that isn't supported or force None sharding on them, as you did in the fwd function.

Reply:
Added.